### ISSUE TITLE
netdata segfault when restart service 

### ISSUE THREAD

**Issue by Wing924** [netdata/netdata#5366]
##### Bug report summary
netdata segfault when restart service (not always)
```
Feb 12 16:36:15 ****** systemd[1]: Stopping Real time performance monitoring...
Feb 12 16:36:22 ****** kernel: [13730643.053967] netdata[14664]: segfault at 7f75baf29159 ip 00007f75bb6b0938 sp 00007f75bb8e5240 error 4 in netdata[7f75bb647000+e6000]
Feb 12 16:36:22 ****** abrt-hook-ccpp[43072]: Process 14636 (netdata) of user 21009 killed by SIGSEGV - dumping core
Feb 12 16:36:22 ****** abrt-hook-ccpp[43072]: Failed to create core_backtrace: waitpid failed: No child processes
Feb 12 16:36:22 ****** systemd[1]: netdata.service: main process exited, code=killed, status=11/SEGV
Feb 12 16:36:22 ****** systemd[1]: Stopped Real time performance monitoring.
Feb 12 16:36:22 ****** systemd[1]: Unit netdata.service entered failed state.
Feb 12 16:36:22 ****** systemd[1]: netdata.service failed.
Feb 12 16:36:22 ****** systemd[1]: Starting Real time performance monitoring...
Feb 12 16:36:22 ****** systemd[1]: Started Real time performance monitoring.
```

##### OS / Environment
```
$ uname -a
Linux ******* 3.10.0-862.11.6.el7.x86_64 #1 SMP Tue Aug 14 21:49:04 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```
##### Netdata version (ouput of `netdata -V`)
```
$ /opt/netdata/bin/netdata -V
netdata 1.11.1_rolling
```
install via static binary.
##### Steps To Reproduce
```
systemctl restart netdata
```

**Comment by cakrit** [comment]
Please [upgrade to the latest netdata](https://docs.netdata.cloud/packaging/installer/update/) and let us know if you still face the issue.

**Comment by Wing924** [comment]
@cakrit I use static binary package. 1.11.1 is the latest released version.

**Comment by cakrit** [comment]
You're right, we will release the binary package today.

**Comment by cakrit** [comment]
https://github.com/netdata/binary-packages/ has been replaced with Google Cloud. We have an issue with `kickstart-static64.sh`. Please monitor  #5390 for the resolution, or [select a different installation method](https://docs.netdata.cloud/packaging/installer/) if it's urgent.

**Comment by Wing924** [comment]
@cakrit I installed latest binary package: v1.12.0.
This still happen to me.

**Comment by cakrit** [comment]
Thanks @Wing924, @mfundul will assist with the analysis. We'll obviously need to get a core dump to investigate.

**Comment by Wing924** [comment]
@cakrit I think it may related to #5367

**Comment by mfundul** [comment]
@Wing924 is this latest CentOS 7 or something else?

**Comment by mfundul** [comment]
Also, how did you get the static binary and how did you install it?

**Comment by Wing924** [comment]
@mfundul 
The results are the same with #5367 (check the matrix I posted)
I just download binary package from release page and run it by hand.

**Comment by mfundul** [comment]
I donwloaded [netdata-v.1.12.0.gz.run](https://github.com/netdata/netdata/releases/download/v1.12.0/netdata-v1.12.0.gz.run). Ran `sudo ./netdata-v1.12.0.gz.run` to install it. OS version from `/etc/redhat-release` is `CentOS Linux release 7.6.1810 (Core)`.

Is this the exact same binary that you're using? Do any of the above steps seem suspicious to you?

**Comment by mfundul** [comment]
In my scenario I cannot reproduce the segfault. Can you gather the core dump for us to examine? Does this [guide](https://iangilham.com/2016/12/08/core-dump-from-centos-7.html) help you?

**Comment by Wing924** [comment]
the os is not centos 7.6 but centos 6.3 or 6.10

**Comment by mfundul** [comment]
Your kernel via `uname -a` says `3.10.0-862.11.6.el7.x86_64`. It is quite impossible to be CentOS 6, firstly because it says `el7` and secondly because CentOS 6 uses kernel 2.6.32 and not kernel 3.10. Please run `cat /etc/redhat-release` to see what the exact version is.

**Comment by Wing924** [comment]
sorry, I wroted the wrong kernel. it’s centos 6, not 7.

**Comment by mfundul** [comment]
Ok, however this issue is very strange, cause there is no systemd or systemctl in CentOS 6, a shown in the original post. In the original post the output was clearly generated from a CentOS 7 system. Issue #5367  seems to make more sense.

In CentOS 6 there is a problem, but not what is described above.

**Comment by Wing924** [comment]
@mfundul sorry, I'm confusing with the other issue. I'll response you next Monday.

**Comment by Wing924** [comment]
@mfundul 
I got coredump. Since it may contains some secret system info, I want to send to you directly.

**Comment by mfundul** [comment]
@Wing924 you can send me a download link to markos@netdata.cloud

**Comment by mfundul** [comment]
@Wing924 thank you for sending me the e-mail. Unfortunately I cannot examine version 1.11.1_rolling cause I do not know the exact git revision of the code and I don't have any debug symbols nor can I produce them.

Can you use version v1.12.0 which is tagged in git and I can debug? And if v1.12.0 does not produce segmentation faults even better!

**Comment by Wing924** [comment]
@mfundul My binary installer is download from github release page (v1.11.1), the `_rolling` suffix is netdata bug. I'm evaluate v1.12.0 now. If I get segmentation faults again, I'll send you.

**Comment by Wing924** [comment]
v1.12.0 (static build) also get segmentation faults. I enabled abrtd. I'll send you coredump next time.

```
# system info
$ ./netdata -v
netdata v1.12.0
$ uname -r
3.10.0-693.11.6.el7.x86_64
$ cat /etc/redhat-release
CentOS Linux release 7.6.1810 (Core)

# syslog
Feb 21 03:26:15 ********* kernel: [24080382.774988] netdata[4149]: segfault at 7f66cedf6159 ip 0000556381325458 sp 00007f66cef63240 error 4 in netdata[5563812a8000+ed000]
Feb 21 03:26:15 ********* abrt-hook-ccpp[60865]: Process 4141 (netdata) of user 21009 killed by SIGSEGV - dumping core
Feb 21 03:26:15 ********* systemd[1]: netdata.service: main process exited, code=killed, status=11/SEGV
Feb 21 03:26:15 ********* systemd[1]: Unit netdata.service entered failed state.
Feb 21 03:26:15 ********* systemd[1]: netdata.service failed.
Feb 21 03:26:15 ********* abrt-server[60866]: Executable '/opt/netdata/bin/srv/netdata' doesn't belong to any package and ProcessUnpackaged is set to 'no'
```

**Comment by mfundul** [comment]
@Wing924 thank you for the core dump you sent. I will post the backtrace for reference:
```
Code = Static binary netdata v.1.12.0
Program terminated with signal 11, Segmentation fault.
#0  rrdset_next_usec.constprop.119 (st=0x7f12dcbd6000, microseconds=0) at database/rrdset.c:749
#1  0x000056220586236e in tc_main (ptr=<optimized out>) at collectors/tc.plugin/plugin_tc.c:1090
#2  0x000056220581787f in thread_start.lto_priv.140 (ptr=0x5622073c4ea0) at libnetdata/threads/threads.c:126
#3  0x00005622058a501a in start (p=0x7f12dd442ee8) at src/thread/pthread_create.c:150
#4  0x00005622058a6095 in __clone () at src/thread/x86_64/clone.s:21
```

**Comment by cakrit** [comment]
The tc plugin hasn't been touched in 4 months though. @Wing924 can you please disable the tc plugin so we can first verify it's not another issue causing the crash? 
In `netdata.conf`...

```
[plugins]
tc = no
```

**Comment by cakrit** [comment]
By the way, we released patch version 1.12.1 yesterday. Perhaps it will resolve your issue, but we're not sure. Can you check that one too? 
So two tests: 
 - 1.12.1 with tc enabled
 - 1.12.0 with tc disabled

**Comment by mfundul** [comment]
There was a second thread during the crash:
```
#0  __wait (addr=addr@entry=0x56220797d44c, waiters=waiters@entry=0x56220797d450, val=60891, priv=128, priv@entry=1)
    at src/thread/__wait.c:14
#1  0x00005622058aa04b in __lockfile (f=f@entry=0x56220797d3c0) at src/stdio/__lockfile.c:10
#2  0x00005622058abdf8 in close_file (f=0x56220797d3c0) at src/stdio/__stdio_exit.c:11
#3  0x00005622058abe43 in __stdio_exit () at src/stdio/__stdio_exit.c:19
#4  0x00005622057f90c8 in exit (code=0) at src/exit/exit.c:32
#5  0x00005622057ff6df in netdata_cleanup_and_exit (ret=ret@entry=0) at daemon/main.c:53
#6  0x00005622057faa9e in signals_handle () at daemon/signals.c:154
#7  main (argc=<optimized out>, argv=<optimized out>) at daemon/main.c:1126
```

**Comment by mfundul** [comment]
@Wing924 can you please post the netdata error log? It should be in `/opt/netdata/var/log/netdata`.

**Comment by cakrit** [comment]
Indeed, we looked at a bit more and all we need is `error.log` right after the crash. Don't bother with the tests I suggested, it's a different issue, at exit.

**Comment by Wing924** [comment]
@cakrit 
this is error.log (v1.12.0 with tc)
```
2019-02-22 03:21:04: netdata INFO  : MAIN : SIGNAL: Received SIGTERM. Cleaning up to exit...
2019-02-22 03:21:04: netdata INFO  : MAIN : EXIT: netdata prepares to exit with code 0...
2019-02-22 03:21:04: netdata INFO  : MAIN : EXIT: cleaning up the database...
2019-02-22 03:21:04: netdata INFO  : MAIN : Cleaning up database [1 hosts(s)]...
2019-02-22 03:21:04: netdata INFO  : MAIN : Cleaning up database of host 'stg-sysdapple103z.stg.jp.local'...
2019-02-22 03:21:04: netdata ERROR : PLUGINSD[apps] : read failed (errno 22, Invalid argument)
2019-02-22 03:21:04: netdata ERROR : PLUGINSD[apps] : '/opt/netdata/usr/libexec/netdata/plugins.d/apps.plugin' (pid 2411) disconnected after 774180 successful data collections (ENDs).
2019-02-22 03:21:04: netdata ERROR : PLUGINSD[apps] : child pid 2411 killed by signal 15.
2019-02-22 03:21:04: netdata ERROR : PLUGINSD[apps] : '/opt/netdata/usr/libexec/netdata/plugins.d/apps.plugin' (pid 2411) exited with error code -1, but has given useful output in the past (4734455 times). Waiting a bit before starting it again. (errno 22, Invalid argument)
2019-02-22 03:21:04: netdata ERROR : PLUGIN[tc] : child pid 23454 killed by signal 15.
2019-02-22 03:21:04: netdata ERROR : PLUGINSD[python.d] : read failed (errno 22, Invalid argument)
2019-02-22 03:21:04: netdata ERROR : PLUGINSD[python.d] : '/opt/netdata/usr/libexec/netdata/plugins.d/python.d.plugin' (pid 60910) disconnected after 1915998 successful data collections (ENDs).
2019-02-22 03:21:04: netdata ERROR : PLUGINSD[python.d] : child pid 60910 killed by signal 15.
2019-02-22 03:21:04: netdata ERROR : PLUGINSD[python.d] : '/opt/netdata/usr/libexec/netdata/plugins.d/python.d.plugin' (pid 60910) exited with error code -1, but has given useful output in the past (1915998 times). Waiting a bit before starting it again. (errno 22, Invalid argument)
2019-02-22 03:21:04: netdata ERROR : PLUGINSD[go.d] : read failed
2019-02-22 03:21:04: netdata ERROR : PLUGINSD[go.d] : '/opt/netdata/usr/libexec/netdata/plugins.d/go.d.plugin' (pid 60908) disconnected after 258258 successful data collections (ENDs). (errno 22, Invalid argument)
2019-02-22 03:21:04: netdata ERROR : PLUGINSD[go.d] : child pid 60908 killed by signal 15.
2019-02-22 03:21:04: netdata ERROR : PLUGINSD[go.d] : '/opt/netdata/usr/libexec/netdata/plugins.d/go.d.plugin' (pid 60908) exited with error code -1, but has given useful output in the past (258258 times). Waiting a bit before starting it again. (errno 22, Invalid argument)
2019-02-22 03:21:06: tc-qos-helper.sh: WARNING: FireQoS is not installed on this system. Use FireQoS to apply traffic QoS and expose the class names to netdata. Check https://github.com/netdata/netdata/tree/master/collectors/tc.plugin#tcplugin
2019-02-22 03:21:06: tc-qos-helper.sh: WARNING: Cannot find file '/opt/netdata/usr/lib/netdata/conf.d/tc-qos-helper.conf'.
2019-02-22 03:21:06: tc-qos-helper.sh: WARNING: Cannot find file '/opt/netdata/etc/netdata/tc-qos-helper.conf'.
2019-02-22 03:21:09: netdata INFO  : MAIN : EXIT: stopping master threads...
2019-02-22 03:21:09: netdata INFO  : MAIN : EXIT: Stopping master thread: PLUGIN[proc]
2019-02-22 03:21:09: netdata INFO  : MAIN : EXIT: Stopping master thread: PLUGIN[diskspace]
2019-02-22 03:21:09: netdata INFO  : MAIN : EXIT: Stopping master thread: PLUGIN[cgroups]
2019-02-22 03:21:09: netdata INFO  : MAIN : EXIT: Stopping master thread: PLUGIN[tc]
2019-02-22 03:21:09: netdata INFO  : MAIN : EXIT: Stopping master thread: PLUGIN[idlejitter]
2019-02-22 03:21:09: netdata INFO  : MAIN : EXIT: Stopping master thread: STATSD
2019-02-22 03:21:09: netdata INFO  : MAIN : EXIT: Stopping master thread: WEB_SERVER[static1]
2019-02-22 03:21:09: netdata INFO  : MAIN : EXIT: Stopping master thread: PLUGINSD
2019-02-22 03:21:09: netdata INFO  : MAIN : EXIT: Stopping master thread: HEALTH
2019-02-22 03:21:09: netdata INFO  : MAIN : Waiting 9 threads to finish...
2019-02-22 03:21:09: netdata INFO  : PLUGIN[idlejitter] : cleaning up...
2019-02-22 03:21:09: netdata INFO  : PLUGIN[idlejitter] : thread with task id 60892 finished
2019-02-22 03:21:09: netdata INFO  : PLUGIN[proc] : cleaning up...
2019-02-22 03:21:09: netdata INFO  : PLUGIN[proc] : thread with task id 60888 finished
2019-02-22 03:21:09: netdata INFO  : PLUGIN[diskspace] : cleaning up...
2019-02-22 03:21:09: netdata INFO  : PLUGIN[diskspace] : thread with task id 60889 finished
2019-02-22 03:21:09: netdata INFO  : PLUGIN[cgroups] : cleaning up...
2019-02-22 03:21:09: netdata INFO  : PLUGIN[cgroups] : thread with task id 60890 finished
2019-02-22 03:21:09: netdata INFO  : STATSD : cleaning up...
2019-02-22 03:21:09: netdata INFO  : STATSD : STATSD: stopping data collection thread 1...
2019-02-22 03:21:09: netdata INFO  : STATSD : STATSD: closing sockets...
2019-02-22 03:21:09: netdata INFO  : STATSD : STATSD: cleanup completed.
2019-02-22 03:21:09: netdata INFO  : STATSD : thread with task id 60893 finished
2019-02-22 03:21:09: netdata INFO  : WEB_SERVER[static1] : freeing local web clients cache...
2019-02-22 03:21:09: netdata INFO  : WEB_SERVER[static1] : stopped after 11670 connects, 11670 disconnects (max concurrent 2), 11766 receptions and 23436 sends
2019-02-22 03:21:09: netdata INFO  : WEB_SERVER[static1] : closing all web server sockets...
2019-02-22 03:21:09: netdata INFO  : WEB_SERVER[static1] : all static web threads stopped.
2019-02-22 03:21:09: netdata INFO  : WEB_SERVER[static1] : thread with task id 60895 finished
2019-02-22 03:21:09: netdata INFO  : PLUGINSD : cleaning up...
2019-02-22 03:21:09: netdata INFO  : PLUGINSD : stopping plugin thread: plugin:apps
2019-02-22 03:21:09: netdata INFO  : PLUGINSD : stopping plugin thread: plugin:python.d
2019-02-22 03:21:09: netdata INFO  : PLUGINSD : stopping plugin thread: plugin:go.d
2019-02-22 03:21:09: netdata INFO  : PLUGINSD : cleanup completed.
2019-02-22 03:21:09: netdata INFO  : PLUGINSD : thread with task id 60896 finished
2019-02-22 03:21:09: netdata INFO  : HEALTH : cleaning up...
2019-02-22 03:21:09: netdata INFO  : HEALTH : thread with task id 60897 finished
2019-02-22 03:21:09: netdata ERROR : STATSD_COLLECTOR[1] : POLLFD: LISTENER: processing POLLNVAL events for slot 0 fd 1 (events = 32, revents = 7) (errno 22, Invalid argument)
2019-02-22 03:21:09: netdata ERROR : STATSD_COLLECTOR[1] : POLLFD: LISTENER: processing POLLNVAL events for slot 1 fd 1 (events = 32, revents = 31)
2019-02-22 03:21:09: netdata INFO  : STATSD_COLLECTOR[1] : cleaning up...
2019-02-22 03:21:09: netdata INFO  : STATSD_COLLECTOR[1] : thread with task id 60911 finished
2019-02-22 03:21:09: netdata INFO  : PLUGINSD[apps] : data collection thread exiting
2019-02-22 03:21:09: netdata INFO  : PLUGINSD[apps] : killing child process pid 2411
2019-02-22 03:21:09: netdata ERROR : PLUGINSD[apps] : Request to kill pid 2411, but it is not running. (errno 3, No such process)
2019-02-22 03:21:09: netdata INFO  : PLUGINSD[apps] : thread with task id 60904 finished
2019-02-22 03:21:09: netdata INFO  : PLUGINSD[python.d] : data collection thread exiting
2019-02-22 03:21:09: netdata INFO  : PLUGINSD[python.d] : killing child process pid 60910
2019-02-22 03:21:09: netdata ERROR : PLUGINSD[python.d] : Request to kill pid 60910, but it is not running. (errno 3, No such process)
2019-02-22 03:21:09: netdata INFO  : PLUGINSD[python.d] : thread with task id 60902 finished
2019-02-22 03:21:09: netdata INFO  : PLUGINSD[go.d] : data collection thread exiting
2019-02-22 03:21:09: netdata INFO  : PLUGINSD[go.d] : killing child process pid 60908
2019-02-22 03:21:09: netdata ERROR : PLUGINSD[go.d] : Request to kill pid 60908, but it is not running. (errno 3, No such process)
2019-02-22 03:21:09: netdata INFO  : PLUGINSD[go.d] : thread with task id 60900 finished
2019-02-22 03:21:09: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:09: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:10: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:10: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:10: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:10: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:10: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:10: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:10: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:10: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:10: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:10: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:11: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:11: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:11: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:11: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:11: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:11: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:11: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:11: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:11: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:11: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:12: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:12: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:12: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:12: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:12: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:12: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:12: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:12: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:12: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:12: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:13: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:13: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:13: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:13: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:13: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:13: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:13: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:13: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:13: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:13: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:14: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:14: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:14: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:14: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:14: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:14: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:14: netdata INFO  : MAIN : Waiting 1 threads to finish...
2019-02-22 03:21:14: netdata ERROR : MAIN : Master thread PLUGIN[tc] takes too long to exit. Giving up... (errno 22, Invalid argument)
2019-02-22 03:21:14: netdata INFO  : MAIN : EXIT: freeing database memory...
2019-02-22 03:21:14: netdata INFO  : MAIN : Freeing all memory for host 'stg-sysdapple103z.stg.jp.local'...
2019-02-22 03:21:14: netdata INFO  : MAIN : EXIT: removing netdata PID file '/run/netdata/netdata.pid'...
2019-02-22 03:21:14: netdata INFO  : MAIN : EXIT: all done - netdata is now exiting - bye bye...
```

**Comment by mfundul** [comment]
I guess it's all down to this line @ktsaou @cakrit:
`2019-02-22 03:21:14: netdata ERROR : MAIN : Master thread PLUGIN[tc] takes too long to exit. Giving up... (errno 22, Invalid argument)`

**Comment by cakrit** [comment]
Somehow the tc plugin must be stuck in `tc_device_commit`, with thread cancelability disabled. It's a huge function (440 lines), with only one debug statement. Lines 985-988 of `collectors/tc.plugin/plugin_tc.c`:

```
                    netdata_thread_disable_cancelability();
                    tc_device_commit(device);
                    // tc_device_free(device);
                    netdata_thread_enable_cancelability();
```

**Comment by cakrit** [comment]
@Wing924, if we create a version of `plugin_tc.c` with a bunch of debug messages, would you be able to build from source? The only other alternative I see is to get your entire ip tables configuration and try to replicate the issue ourselves.

**Comment by Wing924** [comment]
@cakrit 
> if we create a version of `plugin_tc.c` with a bunch of debug messages, would you be able to build from source?

yes, it's ok.

> The only other alternative I see is to get your entire ip tables configuration and try to replicate the issue ourselves.

I disabled iptables service.

**Comment by mfundul** [comment]
I think the issue is somewhat different @cakrit. `collectors/tc.plugin/plugin_tc.c` in function `tc_main_cleanup()` does this ` static_thread->enabled = NETDATA_MAIN_THREAD_EXITED;` at LOC 855.

However, if the `tc_main` thread is successfully killed, then there will never be a chance to call `netdata_thread_cleanup_push()` that calls  `tc_main_cleanup()` and `NETDATA_MAIN_THREAD_EXITED` will never be set.

Since `NETDATA_MAIN_THREAD_EXITED` was never set, the lines:
```
if (static_threads[i].enabled != NETDATA_MAIN_THREAD_EXITED)
    found++;
```
in `daemon/main.c` LOC 207 will lead to a timeout.

**Comment by cakrit** [comment]
Actually, you're right. The SIGTERM kills apps, tc, python.d and go.d before the thread cleanup is even called. Notice that when PLUGINSD tries to clean up the external plugins, it complains about the external plugins apps, python.d and go.d, because they're already gone, unexpectedly so. 

This isn't limited to just these plugins. Every plugin that runs as its own process is immediately killed by a systemd restart of netdata. In my case for example, I also have freeipmi, which is also killed immediately. 

We'll need to discuss again with @ktsaou. @Wing924 we appreciate your help, we won't need anything more from you.

**Comment by cakrit** [comment]
Actually, no, it's a bit more complicated. tc itself is a thread running from inside netdata. It's not what gets killed. It's the tc-qos-helper command that is constantly running as a separate process that gets killed. You can see at 2019-02-22 03:21:06 the three warnings that we get when that process is restarted. I'll debug a bit more...

**Comment by mfundul** [comment]
I have built three candidate binaries that are patched in the following way:
* patch 1
```
--- a/libnetdata/threads/threads.c
+++ b/libnetdata/threads/threads.c
@@ -115,8 +115,8 @@ static void *thread_start(void *ptr) {
     if(!(netdata_thread->options & NETDATA_THREAD_OPTION_DONT_LOG_STARTUP))
         info("thread created with task id %d", gettid());
 
-    if(pthread_setcanceltype(PTHREAD_CANCEL_DEFERRED, NULL) != 0)
-        error("cannot set pthread cancel type to DEFERRED.");
+    if(pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, NULL) != 0)
+        error("cannot set pthread cancel type to ASYNCHRONOUS.");
 
     if(pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL) != 0)
         error("cannot set pthread cancel state to ENABLE.");
```
* patch 2
```
--- a/collectors/tc.plugin/plugin_tc.c
+++ b/collectors/tc.plugin/plugin_tc.c
@@ -982,10 +982,8 @@ void *tc_main(void *ptr) {
                 // debug(D_TC_LOOP, "END line");
 
                 if(likely(device)) {
-                    netdata_thread_disable_cancelability();
                     tc_device_commit(device);
                     // tc_device_free(device);
-                    netdata_thread_enable_cancelability();
                 }
 
                 device = NULL;
```
* patch 3
```
diff --git a/daemon/main.c b/daemon/main.c
index e93e97ca..d0afe215 100644
--- a/daemon/main.c
+++ b/daemon/main.c
@@ -197,7 +197,7 @@ void cancel_main_threads() {
             found++;
         }
     }
+    netdata_exit = 1;
     while(found && max > 0) {
         max -= step;
         info("Waiting %d threads to finish...", found);
```

@Wing924 can you see if this fixes your problem? Ideally the first patch should be enough, hopefully you don't have to go through all of them.

**Comment by Wing924** [comment]
> By the way, we released patch version 1.12.1 yesterday. Perhaps it will resolve your issue, but we're not sure. Can you check that one too?
> So two tests:
> 
> * 1.12.1 with tc enabled
> * 1.12.0 with tc disabled

1.12.1 with tc enabled => still happen
1.12.0 with tc disabled => OK

**Comment by Wing924** [comment]
@mfundul I applied 3 patches and they look OK for now but I need more time to check if it's be fixed. 
btw, don't you think use `PTHREAD_CANCEL_ASYNCHRONOUS` is too dangerous?

**Comment by mfundul** [comment]
@Wing924 did you apply all 3 patches at the same time? It is possible that you won't see the problem without a static binary. What command did you use to install the patches? You can test individual static binaries from the google drive link I posted above.

If you prefer to build the static binaries yourself, you need to run the command `./.travis/create_artifacts.sh` after applying each patch.

About `PTHREAD_CANCEL_ASYNCHRONOUS` I don't see a danger because the code has been architected in a way that it can be interrupted at any point except for a few selected regions where cancelability is disabled. About the tc plugin, we identified this region is not necessary.

**Comment by Wing924** [comment]
@mfundul 
> did you apply all 3 patches at the same time?

I run the 3 binary you provided on 3 different servers. Now they all look OK.
And on another server, I installed netdata v1.12.1 from source (no patch), which also looks OK.

I'm not expert of linux kernel, I just googled some about `PTHREAD_CANCEL_ASYNCHRONOUS` and find:
> Setting the cancelability type to PTHREAD_CANCEL_ASYNCHRONOUS is rarely useful. Since the thread could be canceled at any time, it cannot safely reserve resources (e.g., allocating memory with malloc(3)), acquire mutexes, semaphores, or locks, and so on.

https://linux.die.net/man/3/pthread_setcancelstate

**Comment by mfundul** [comment]
> 
> 
> @mfundul
> 
> > did you apply all 3 patches at the same time?
> 
> I run the 3 binary you provided on 3 different servers. Now they all look OK.
> And on another server, I installed netdata v1.12.1 from source (no patch), which also looks OK.
> 
> I'm not expert of linux kernel, I just googled some about `PTHREAD_CANCEL_ASYNCHRONOUS` and find:
> 
> > Setting the cancelability type to PTHREAD_CANCEL_ASYNCHRONOUS is rarely useful. Since the thread could be canceled at any time, it cannot safely reserve resources (e.g., allocating memory with malloc(3)), acquire mutexes, semaphores, or locks, and so on.
> 
> https://linux.die.net/man/3/pthread_setcancelstate
@Wing924 you are right that `PTHREAD_CANCEL_ASYNCHRONOUS` is the most risky of the 3 patches, meaning that it is going to require much more testing and more careful code development from now on. 

For that reason, if the other 2 patches prove to be adequate it would be wiser to avoid patch 1 to minimize risk.

Furthermore, patch 3 is code we're trying to phase out, eventually we want to remove the `netdata_exit` variable from this project.

**Comment by Wing924** [comment]
@mfundul 
I got SIGSEGV on patch 2, patch 1 & 3 are OK.
I wonder why compiling from source without patch works well but static binary gets SIGSEGV?

**Comment by mfundul** [comment]
> 
> 
> @mfundul
> I got SIGSEGV on patch 2, patch 1 & 3 are OK.
> I wonder why compiling from source without patch works well but static binary gets SIGSEGV?

The difference is that static binaries need to use musl libc whereas compiling from source uses glibc. Canceling pthreads uses a badly designed POSIX API and its behaviour is very implementation specific.

**Comment by Wing924** [comment]
> > @mfundul
> > I got SIGSEGV on patch 2, patch 1 & 3 are OK.
> > I wonder why compiling from source without patch works well but static binary gets SIGSEGV?
> 
> The difference is that static binaries need to use musl libc whereas compiling from source uses glibc. Canceling pthreads uses a badly designed POSIX API and its behaviour is very implementation specific.

This is one reason. I think the other reason is netdata heavily use of `#ifdef` so that compiling can choose right implementation, but static binary can't.

I think we should not use thread cannel because of it's badly designed POSIX API, instead, make thread function exit.
Many other language don't support or deprecated cancel threads:
* Java: `Thread.stop()` is deprecated because it's dangerous.
* Go: don't support cannel, user need to exit thread function.

**Comment by mfundul** [comment]
> 
> 
> > > @mfundul
> > > I got SIGSEGV on patch 2, patch 1 & 3 are OK.
> > > I wonder why compiling from source without patch works well but static binary gets SIGSEGV?
> > 
> > 
> > The difference is that static binaries need to use musl libc whereas compiling from source uses glibc. Canceling pthreads uses a badly designed POSIX API and its behaviour is very implementation specific.
> 
> This is one reason. I think the other reason is netdata heavily use of `#ifdef` so that compiling can choose right implementation, but static binary can't.
> 
> I think we should not use thread cannel because of it's badly designed POSIX API, instead, make thread function exit.
> Many other language don't support or deprecated cancel threads:
> 
>     * Java: `Thread.stop()` is deprecated because it's dangerous.
> 
>     * Go: don't support cannel, user need to exit thread function.

The `#ifdef`s have been a cause for many bugs for the static binary, that is absolutely correct. In this case, I think it's not related thanks to your core dump. Eventually, we will move to `libuv` and we won't be canceling threads this way.
